feat: uCOS3 port for Sifive E2x RISC-V

iontra-gaurav opened the pull request:
## Description
uCOS3 port for Sifive E2x RISC-V

This commit contains changes for enabling interrupt(including timer interrupt) in uC-OS RISC V port.

- Saving and restoring of registers during context switch (task to task, interrupt to task, task to interrupt)
- Removed systick handler as we are directly calling OSTimeTick() from timer ISR.
- Added comments in os_cpu_a.S file for better understanding of assembly line instruction.

## Jira Links

## Type of change
* New feature (non-breaking change which adds functionality)


## How Has This Been Tested?
This change has been tested with osa_task_switch test case. and it supposed be to switch task to task,task to interrupt 
and interrupt to task, during any transition it should save the current context and load the next context. 

## Checklist:

- [x] Comments updated and spell checked
- [x] Corresponding documentation changes included
- [x] Review commits squashed
- [x] Final commit retested

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -78,24 +83,34 @@
 
 OSStartHighRdy:
 # Disable interrupts
+    # load immediately "RISCV_MSTATUS_MIE" into temp to register
```
into temp to register -> into t0 register

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -78,24 +83,34 @@
 # Disable interrupts
+    # load immediately "RISCV_MSTATUS_MIE" into temp to register
     li     t0, RISCV_MSTATUS_MIE
+    # mstatus = mstatus & t0, no affect on t0 register 
```
No need to say this "no affect on t0 register"

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -78,24 +83,34 @@
 
 # SWITCH TO HIGHEST PRIORITY TASK
     la     t0, OSTCBHighRdyPtr
+    #load word from address stored in [t0 + 0(offset)] to t1
```
#load -> # load

Similarly in order places

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +169,92 @@ OSStartHighRdy:
+    ecall
     ret
 
+.align 8
```
Put a comment saying in RISCV exception handler MUST be 8 byte aligned

kedar-chitnis replied:
Sorry, I think align 8 aligns to 1<<8 i.e 256 byte so change the comment to say "256 bytes"

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +169,92 @@ OSStartHighRdy:
+    csrr  t0, mcause
+    and t3, t1, t0
+    li t1, 11
+  	csrr   t0,  mepc
```
Check the tab and align to previous line

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +169,92 @@ OSStartHighRdy:
+
+
+ 
+    li t1, ARCH_CPU_MCAUSE_CAUSE_MASK
```
Put comment here, basically if its a interrupt we dont add +4 and if its a exception we add +4 to mepc

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +169,92 @@ OSStartHighRdy:
+    bne t1, t3, DONTADD
+    addi   t0,  t0, 4  //it will jump just after ecall .i.e ecall + 4
+   
+DONTADD: # for interrupt we dont need to add +4 in sp
```
dont need to add +4 in mepc

kedar-chitnis replied:
This comment is still open looks like

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +169,92 @@ OSStartHighRdy:
+DONTADD: # for interrupt we dont need to add +4 in sp
+    sw     t0,  31*4(sp)
+
+    la     t0, OSTCBCurPtr
```
Put a comment saying what this does
OSTCBCurPtr = SP

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -262,9 +337,11 @@ Software_IRQHandler:
 
 # Exception return will restore remaining context
+    
+    li t0, 0x80 #set MPIE = 1
```
Put some more comment here like
setting MPIE = 1 will make sure when mret is called, interrupts are enabled

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -78,27 +83,38 @@
 # Disable interrupts
+    # load immediately "RISCV_MSTATUS_MIE" into t0 register
     li     t0, RISCV_MSTATUS_MIE
+    # mstatus = mstatus & t0
```
This operation is more like "mstatus &= ~t0" , i.e it clears bits which are set in t0

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -55,6 +56,10 @@
 
     .equ  RISCV_PRCI_BASE_ADDR,      0x44000000
 
+    .equ  ARCH_CPU_MCAUSE_INTR_MASK,  0x80000000
```
RISCV_PRCI_BASE_ADDR, RISCV_MIE_MSIE is not used anywhere, remove it

kedar-chitnis replied:
RISCV_MIE_MSIE can be removed, its not used

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -43,6 +43,7 @@
     .global  OSCtxSw
     .global  OSIntCtxSw
     .global  Software_IRQHandler
+    .global  ucos_intr_exception_handler
```
OSIntExit is not used anywhere remove it

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -43,6 +43,7 @@
     .global  OSCtxSw
```
There is this line at the top
"Hardware FP is not supported."

Later when we add floating point support, make sure to remove this line

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +170,94 @@ OSStartHighRdy:
+
+    # If it is a interrupt, do not add 4 in mepc
+    # If it is a exception, add 4 in mepc
+    li t1, ARCH_CPU_MCAUSE_CAUSE_MASK
```
You need to check for interrupt or exception here right, i.e ARCH_CPU_MCAUSE_INTR_MASK

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,19 +170,94 @@ OSStartHighRdy:
+    li t1, ARCH_CPU_MCAUSE_INTR_MASK
+    csrr  t0, mcause
+    and t2, t1, t0
+    li t1, 1
```
instead of 1, shouldnt this be ARCH_CPU_MCAUSE_INTR_MASK

kedar-chitnis replied:
And t1 is already loaded on top with ARCH_CPU_MCAUSE_INTR_MASK

 so maybe this line is not needed

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -193,20 +285,6 @@ OSIntCtxSw:
 #********************************************************************************************************
 
 Software_IRQHandler:
```
Check the comments at top of this function and change the words to reflect the changes we did

I think remove
all of 1), 
2.a,b,c 
all of 3)

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -262,9 +340,12 @@ Software_IRQHandler:
 
 # Exception return will restore remaining context
+    # set MPIE = 1
+    # interrupts will be enabled from here
```
The correct comment is 
interrupts will be enabled after mret is called

kedar-chitnis commented on Ports/RISC-V/RV32/GCC/os_cpu_a.S:
```diff
@@ -153,60 +168,109 @@ OSStartHighRdy:
 #              the result of an interrupt.  This function triggers a synchronous software interrupt by
 #              writing into the MSIP register
 #********************************************************************************************************
+OSIntCtxSw:
```
Check the comments above this line, and change them to match the actual code.
e.g we use ecall and not MSIP

Also we keep OSIntCtxSw as blank since the context switch action after interrupt is done just before exiting the interrupt handler

iontra-gaurav committed:
feat: uCOS3 port for Sifive E2x RISC-V

-This commit contains changes for enabling interrupt(including timer interrupt) in uC-OS RISC V port.
-Saving and restoring of registers during context switch (task to task, interrupt to task, task to interrupt)
-Removed systick handler as we are directly calling OSTimeTick() from timer ISR.
-Added comments in os_cpu_a.S file for better understanding of assembly line instruction.

Signed-off-by: Gaurav <gaurav@iontra.com>

kedar-chitnis commented:
Change the commit message to say "uCOS3 port for Sifive E2x RISC-V"

Also update the PR header and description to match the commit message
